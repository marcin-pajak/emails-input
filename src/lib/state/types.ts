import { ActionType, LibraryState } from './constants';

/**
 * State Actions
 */
export interface AddEmailAction {
  payload: string;
  type: typeof ActionType.Add;
}

export interface RemoveEmailAction {
  payload: string;
  type: typeof ActionType.Remove;
}

export interface UpdateStateAction {
  payload: LibraryState;
  type: typeof ActionType.UpdateState;
}

export type StateAction =
  | AddEmailAction
  | RemoveEmailAction
  | UpdateStateAction;

/**
 * State
 */
export interface Entry {
  email: string;
  isValid: boolean;
}

export type LocalId = string;

export type EmailMapper = Record<string, LocalId>;

export type EntryMap = Record<string, Entry>;

export interface State {
  emails: string[];
  entries: Entry[];
  entryMap: EntryMap;
  mapper: EmailMapper;
  state: LibraryState;
}

/**
 * State Manager
 */
export interface StateManager {
  handleAction: (action: StateAction, state: State) => string | undefined;
  makeState: (initialValue: string[]) => State;
}
