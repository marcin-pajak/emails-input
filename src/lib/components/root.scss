.EmailsInput {
  align-content: flex-start;
  align-items: flex-start;
  background: #fff;
  border: 1px solid #c3c2cf;
  border-radius: 4px;
  display: flex;
  flex-wrap: wrap;
  font-family: 'Open Sans', sans-serif;
  height: inherit;
  max-height: inherit;
  overflow-x: hidden;
  overflow-y: auto;
  padding: 2px;
  word-break: break-word;
  width: 100%;
}

@supports (-moz-appearance: none) {
  .EmailsInput::after {
    content: '';
    display: block;
    height: 2px;
    width: 100%;
    opacity: 0;
  }
}
