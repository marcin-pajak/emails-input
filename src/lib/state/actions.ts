import { isValidEmail } from '../../utils/isValidEmail';
import {
  AddEmailAction,
  RemoveEmailAction,
  State,
  UpdateStateAction,
} from './types';

/**
 * Remove email
 * @param action
 * @param state
 */
export const handleRemoveAction = (
  action: RemoveEmailAction,
  state: State,
): void => {
  const name = state.entryMap[action.payload].email;
  const index = state.emails.indexOf(name);
  delete state.entryMap[action.payload];
  delete state.mapper[name];
  state.entries.splice(index, 1);
  state.emails.splice(index, 1);
};

/**
 * Add email
 * Returns local Id of new email
 * @param action
 * @param state
 * @param newId
 */
export const handleAddAction = (
  action: AddEmailAction,
  state: State,
  newId: string,
): void => {
  const entry = {
    email: action.payload,
    isValid: isValidEmail(action.payload),
  };
  state.emails.push(action.payload);
  state.mapper[action.payload] = newId;
  state.entryMap[newId] = entry;
  state.entries.push(entry);
};

/**
 * Update Library state
 * @param action
 * @param state
 */
export const handleUpdateAction = (
  action: UpdateStateAction,
  state: State,
): void => {
  state.state = action.payload;
};
