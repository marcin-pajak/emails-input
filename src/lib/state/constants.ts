export enum ActionType {
  Add = 'ADD',
  Remove = 'REMOVE',
  Initialize = 'INITIALIZE',
  Destroy = 'DESTROY',
  UpdateState = 'UPDATE_STATE',
}

export enum LibraryState {
  Initializing = 'Initializing',
  Initialized = 'Initialized',
  Destroyed = 'Destroyed',
}
