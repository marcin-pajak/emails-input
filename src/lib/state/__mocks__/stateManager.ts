import { LibraryState } from '../constants';
import { State, StateManager } from '../types';

export const stateManager = (): StateManager => ({
  handleAction: jest.fn(),
  makeState: (): State => ({
    emails: ['new@email.com'],
    entries: [
      {
        email: 'new@email.com',
        isValid: true,
      },
    ],
    entryMap: {
      '1': {
        email: 'new@email.com',
        isValid: true,
      },
    },
    mapper: {
      'new@email.com': '1',
    },
    state: LibraryState.Initialized,
  }),
});
