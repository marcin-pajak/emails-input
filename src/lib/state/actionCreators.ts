import { ActionType, LibraryState } from './constants';
import { AddEmailAction, RemoveEmailAction, UpdateStateAction } from './types';

/**
 * Add new email
 * @param email
 */
export const addItemAction = (email: string): AddEmailAction => ({
  payload: email,
  type: ActionType.Add,
});

/**
 * Remove email
 * @param id
 */
export const removeItemAction = (id: string): RemoveEmailAction => ({
  payload: id,
  type: ActionType.Remove,
});

/**
 * Update Library's state
 * @param state
 */
export const updateStateAction = (state: LibraryState): UpdateStateAction => ({
  payload: state,
  type: ActionType.UpdateState,
});
