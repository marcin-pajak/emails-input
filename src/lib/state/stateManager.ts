import { isValidEmail } from '../../utils/isValidEmail';
import { objectValues } from '../../utils/object';
import {
  handleAddAction,
  handleRemoveAction,
  handleUpdateAction,
} from './actions';
import { ActionType, LibraryState } from './constants';
import {
  EmailMapper,
  Entry,
  EntryMap,
  State,
  StateAction,
  StateManager,
} from './types';

/**
 * Library's State Manipulation
 */
export const stateManager = (): StateManager => {
  const id = {
    current: 1,
  };

  /**
   * Return new state
   * @param initialValue
   */
  const makeState = (initialValue: string[]): State => {
    const mapper: EmailMapper = {};

    const entryMap = initialValue.reduce(
      (map: EntryMap, email: string): EntryMap => {
        const newId = String(id.current);
        map[newId] = {
          email,
          isValid: isValidEmail(email),
        };
        mapper[email] = newId;
        id.current++;

        return map;
      },
      {},
    );

    return {
      emails: initialValue,
      entries: objectValues<Entry>(entryMap),
      entryMap,
      mapper,
      state: LibraryState.Initializing,
    };
  };

  /**
   * Process action
   * @param action
   * @param state
   */
  const handleAction = (
    action: StateAction,
    state: State,
  ): string | undefined => {
    switch (action.type) {
      case ActionType.Add: {
        const newId = String(id.current);
        handleAddAction(action, state, newId);
        id.current++;

        return newId;
      }

      case ActionType.Remove: {
        handleRemoveAction(action, state);

        return undefined;
      }

      case ActionType.UpdateState: {
        handleUpdateAction(action, state);

        return undefined;
      }

      default: {
        return undefined;
      }
    }
  };

  return {
    handleAction,
    makeState,
  };
};
