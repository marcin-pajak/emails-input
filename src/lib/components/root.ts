import { CLASSES } from '../constants';
import { State } from '../state/types';
import { emailsTemplate } from './emails';
import { inputTemplate } from './input';

/**
 * Root Component Template
 * @param state
 */
export const rootTemplate = (state: State): string => `
  <div class="${CLASSES.WRAPPER}">
    ${emailsTemplate(state.entryMap)}
    ${inputTemplate()}
  </div>
`;
