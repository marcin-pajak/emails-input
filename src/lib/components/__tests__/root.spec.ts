import { stateManager } from '../../state/stateManager';
import { rootTemplate } from '../root';

jest.mock('../../state/stateManager');
jest.mock('../emails');
jest.mock('../input');

describe('root template should', (): void => {
  it('return valid template', (): void => {
    expect.hasAssertions();

    const state = stateManager().makeState([]);

    expect(rootTemplate(state)).toMatchSnapshot();
  });
});
