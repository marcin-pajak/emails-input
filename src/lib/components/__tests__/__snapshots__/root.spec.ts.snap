// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`root template should return valid template 1`] = `
"
  <div class=\\"EmailsInput\\">
    <div />
    <input />
  </div>
"
`;
